madgraph5: add newer versions and a pythia8 variant

jmcarcell opened the pull request:
EDIT:
- Use the github url to fetch experimental versions since the launchpad ones were given problems, see [Wouter's comment](https://github.com/spack/spack/pull/41128#issuecomment-1816540012) and [Olivier's comment](https://github.com/spack/spack/pull/41128#issuecomment-2077343326)
- Set the LTS versions as preferred while having other older LTS (deprecated) and newer experimental versions from Github
- Add a pythia8 variant to support both the installation of the pythia8 interface and setting the path for pythia8 in madgraph. Then, one can specify `shower=pythia` and madgraph will generate the commands and run them.
- Remove building `StdHEP` since this has been deprecated for a while, see https://github.com/mg5amcnlo/mg5amcnlo/issues/95#issuecomment-2077464163
- Add a patch to be able to run from CVMFS on AFS, see https://github.com/mg5amcnlo/mg5amcnlo/pull/96

wdconinc commented:
There are some problems with these launchpad links. They stop working or they renamed the url directory from 2.0 to lts or they remove older versions.  E.g. https://launchpad.net/mg5amcnlo/2.0/2.8.x/+download/MG5_aMC_v2.8.1.tar.gz explicitly listed for 2.8.1 is not working, neither is automatically derived https://launchpad.net/mg5amcnlo/2.0/2.8.x/+download/MG5_aMC_v2.8.3.tar.gz.

Mg5 is now also releasing on GitHub which is going to be a bit more stable. Does it make sense to start using that? E.g. https://github.com/mg5amcnlo/mg5amcnlo/archive/refs/tags/v3.5.2.tar.gz

jmcarcell commented:
I did what you mention @wdconinc about the urls (indeed they seem to work at some point in time and not work later, as if they were deleted) and added a few other changes, see the top post.

jmcarcell committed:
madgraph5amc: Add newer versions and a pythia8 variant

wdconinc commented:
Can you comment on the need for a new patch? Is there an upstream bug report?

jmcarcell commented:
This is the report with all the information about the patch: https://github.com/mg5amcnlo/mg5amcnlo/issues/95

jmcarcell committed:
Fix style

oliviermattelaer commented:
Hi,

Note that the github tarball are development version and are different than the tarball distributed on launchpad.
madgraph is actually using github but release and user support is still done trough launchpad.
Using the github tarball will deactivate some feature, activate debug mode,...
So I would not recommend it.

In term of tarball only tarball for LTS are guarantee to stay online for long period, since indeed other tarball are automatically removed when new bug fixing version is released.

wdconinc commented:
@jmcarcell I'm going to let you decide how to handle the versioning, since my community is not a heavy user of mg5. With my spack maintainer hat on, I don't think it is a good user experience for spack users when they suddenly cannot download a version that is listed in the package, so @oliviermattelaer's comment would then indicate that versions may need to be limited to LTS versions from launchpad only (after they have been renamed to lts). Similarly, spack users often run on HPC systems where performance is important, so running in debug mode is not desirable for those users. Then again, others use spack to explore the compatibility of new versions in software stacks. Maybe a middle road is the following:
- most recent LTS is `preferred=True` version,
- newer non-LTS versions (which may not have valid links in the future) are not going to be installed by default,
- older non-LTS versions are marked as `deprecated=True` when the LTS is released, so people who have pinned them in their environments will get notified on concretization, not only on installation.
- deprecated versions are removed in the next spack version.

jmcarcell committed:
Update madgraph5amc versions, deprecated some, add a patch

and don't build StdHEP (deprecated according to the authors)

jmcarcell committed:
Remove urls

jmcarcell committed:
Add ?full_index=1

jmcarcell commented:
I have updated a few things that I edited in the top post:
- Added back a preferred LTS version, while keeping 3.5.4 from github (because since it's not LTS it will disappear from launchpad), and deprecated older ones from launchpad.
- Removed building `StdHEP`
- Added a patch to run from CVMFS on AFS that was found last week. It remains to test this, which I will do in the next days

jmcarcell commented:
I think this PR is ready to be merged if there aren't any more comments.

wdconinc commented on var/spack/repos/builtin/packages/madgraph5amc/package.py:
```diff
@@ -46,15 +63,20 @@ class Madgraph5amc(MakefilePackage):
+    patch(
+        "https://patch-diff.githubusercontent.com/raw/mg5amcnlo/mg5amcnlo/pull/96.diff?full_index=1",
+        sha256="ac6644f1d0ef51d9bdb27a1519261f1cf27d075d39faa278fbc2849acbc5575d",
+    )
```
```suggestion
        when="@3:3.5",
    )
```

wdconinc commented on var/spack/repos/builtin/packages/madgraph5amc/package.py:
```diff
@@ -120,3 +136,13 @@ def installfile(filename):
             join_path(prefix, "Template", "LO", "Source", "make_opts"),
         )
+
+        if "pythia8" in spec:
```
```suggestion
        if "+pythia8" in spec:
```

wdconinc committed:
madgraph5amc: fix typo +pythia8

oliviermattelaer commented:
Note that 3.5.5 is now released and has a status of LTS version too.
(and that 2.9.20 is also released -- the previous LTS-- which will not have support anymore in 2026).

Would not make sense to target those two releases? (and therefore both use the proer launchpad tarball which will be stable)

jmcarcell commented:
@oliviermattelaer I think there is a mistake on the URL for 2.9.20: It's https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v2.9.20.tar.gz when it should be https://launchpad.net/mg5amcnlo/2.0/2.9.x/+download/MG5_aMC_v2.9.20.tar.gz

jmcarcell committed:
Add version range for the AFS patch

Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>

oliviermattelaer commented:
This is the link that you are looking for (I guess):
https://launchpad.net/mg5amcnlo/lts/2.9.x/+download/MG5_aMC_v2.9.20.tar.gz

jmcarcell committed:
Add the LTS versions 3.5.5 and 2.9.20; deprecate 2.9.19

jmcarcell commented:
Yes but so far we were using links like this one: https://launchpad.net/mg5amcnlo/2.0/2.9.x/+download/MG5_aMC_v2.9.19.tar.gz. which were working fine until 2.9.19. I can just change to the LTS one if that's the one we should be using.

jmcarcell commented:
Although then some versions that are in the recipe can't be downloaded anymore and that would be a breaking change for people that are using the recipe now. Can we keep using links like the one I mentioned above that work until 2.9.19? The mistake I mention (2.9.19 being under 3.0/3.5.x) is accessible from the main launchpad page, under downloads both links point to 3.0/3.5.x/...

jmcarcell committed:
Add missing hash

wdconinc commented:
@oliviermattelaer Distributions (not only spack) can only package versions that are provided under a stable link. Can mg5nlo provide such stable links for non-lts versions?

oliviermattelaer commented:
Not for non LTS no. For spack, I would strongly suggest to stick with LTS release.  
If needed, I can provide a "latest" link that will always point to the latest non LTS release. Would you be interested in that?

wdconinc commented:
> Would you be interested in that?

No, as it would be a url that would again change itsl hash unpredictably.

jmcarcell committed:
Fix style

oliviermattelaer commented:
Concerning the change of path for the 2.9.x download links.  The change of download path was just a un-expected side effect of a new presentation of our code in launchpad. We did change the name of a "series" in launchpad,
and looks like this translate to a "mv" and not to a "ln -s" on the system (this is ubuntu decision obviously).
This is/was a "one time change" with the goal to improve the readability of the development structure of our release)
Maybe it was a bad move, but moving back again is certainly now a bad move too.

Now I know that we are aggressive with our short time availability releases and that this is annoying for various project --like spack--. This is actually one of the reason why we now have a LTS type of release such that you have the choice to either use the latest developer version or use a LTS one.

But if you really insist on offering "reproducible" version that we do not wish to continue to distribute within our website,
I can propose one easy solution:
 -  Use github tarball
 -  make an alias of the binary such that the binary is called via python -O $MG5PATH/bin/mg5_aMC
 
Note that If you want full reproducibility, it would be better to remove the use of the "install mg5amc_pythia8_interface" in the spack formula since this command is going to download the latest version of the HepToolsInstaller package, which will download the lastest version of the mg5ac_pythia8_interface and install it. Not really what you want, I guess.
(For that code, the only handle is the hash of the github repo)

jmcarcell committed:
Deprecate another version and add an url_for_version

jmcarcell committed:
Fix url getter

jmcarcell committed:
Remove unnecessary patch

jmcarcell commented:
I propose to keep this PR as is, deprecating all the non-LTS versions before version 3. Then, for the next round of changes we can remove those and move to Github and also deal with the pythia interface.  Removing old versions is also necessary if we don't want to hardcode all the versions of the deprecated versions, since the releases on Github don't reach as back.

bernhardkaindl committed:
Fix urls of older LTS versions, extend timeout

Signed-off-by: Bernhard Kaindl <bernhardkaindl7@gmail.com>

bernhardkaindl reviewed (approved):
@jmcarcell, @wdconinc Further fixes applied:

The URLs of the older LTS versions have changed to /lts/ URLs, and only the last versions are still preserved.

Launchpad repeatedly exceeded the 10 seconds default download timeout of spack, so I added a default timeout of 60 seconds for these downloads.

I merged the changes of https://github.com/spack/spack/pull/46008 to fix the build of 3.5.5 with gcc-14 and array-bounds.patch can only be applied up to 2.8.1, not above.

Tested the build of all versions.

With this, this PR is finally complete and can be merged.
